src/ionosphere/waccmx/edyn_grid_comp.F90

```diff
@@ -0,0 +1,390 @@
+      integer                       :: petCount
+      integer                       :: rc
+      type(ESMF_VM)                 :: vm_init
+      character(len=*), parameter   :: subname = 'edyn_grid_comp_init'
+
+      call edyn_esmf_init(mpi_comm)
+
+      !! Gather PE information for this instance
+      call ESMF_VMGetCurrent(vm_init, rc=rc)
+      call edyn_esmf_chkerr(subname, 'ESMF_VMGetCurrent', rc)
+      call ESMF_VMGet(vm_init, localPet=localPet, petCount=petCount)
+      call edyn_esmf_chkerr(subname, 'ESMF_VMGet', rc)
+      call mpi_comm_size(mpi_comm, npes, rc)
+      call mpi_comm_rank(mpi_comm, iam, rc)
+      ! Collect all the PETS for each instance for phys grid
+      allocate(petlist(npes))
+      call mpi_allgather(localPet, 1, MPI_INTEGER, petlist, 1, &
+           MPI_INTEGER, mpi_comm, rc)
+      ! Now, we should be able to create a gridded component
+      phys_comp = ESMF_GridCompCreate(name=trim(inst_name), petList=petlist,  &
+           rc=rc)
+      call edyn_esmf_chkerr(subname, 'ESMF_GridCompCreate '//trim(inst_name), &
+           rc)
+      call ESMF_GridCompSetServices(phys_comp, edyn_gcomp_SetServices, rc=rc)
+      call edyn_esmf_chkerr(subname,                                          &
+           'ESMF_GridCompSetServices '//trim(inst_name), rc)
+      ! Initialize the required component arguments
+      gcomp_state = ESMF_StateCreate(name='gcomp state', rc=rc)
+      call edyn_esmf_chkerr(subname, 'ESMF_StateCreate', rc)
+      call ESMF_GridCompInitialize(phys_comp, importState=gcomp_state,        &
+           exportState=gcomp_state, rc=rc)
+      call edyn_esmf_chkerr(subname, 'ESMF_GridCompInitialize', rc)
+
+   end subroutine edyn_grid_comp_init
+
+   subroutine edyn_grid_comp_run1(ionos_epotential_model_in,                  &
+        cols, cole, efx_phys, kev_phys)
+
+      use ESMF,         only: ESMF_GridCompRun
+      use edyn_esmf,    only: edyn_esmf_chkerr
+
+      ! Dummy arguments
+      character(len=*),           intent(in) :: ionos_epotential_model_in
+      integer,          optional, intent(in) :: cols
+      integer,          optional, intent(in) :: cole
+      real(r8),         optional, pointer    :: efx_phys(:)
+      real(r8),         optional, pointer    :: kev_phys(:)
+      ! Local variables
+      integer                                :: rc
+      character(len=*), parameter            :: subname = 'edyn_grid_comp_run1'
+
+      do_run = 1
+      if ( present(cols) .and. present(cole) .and. present(efx_phys) .and.    &
+           present(kev_phys)) then
+         amie_efxg => efx_phys
+         amie_kevg => kev_phys
+         col_start = cols
+         col_end = cole
+      else
+         ! No else check assume no optional arguments are passed
+         nullify(amie_efxg)
+         nullify(amie_kevg)
+      end if
+      ionos_epotential_model = ionos_epotential_model_in
+      call ESMF_GridCompRun(phys_comp, importState=gcomp_state,               &
```

Comment:
Are the importState and exportState needed?  You don't seem to do anything with them in the _run routine.  As far as I can tell gcomp_state does not have any fields.

Reply:
Right. I ran into Rocky and Bob in the hall and they suggested a workaround using pointers (see comment below). I guess they are optional arguments in the calls (but need to be dummy arguments in the gridded component interface routines).